### PR TITLE
allow to configure bug/review properties per project

### PR DESCRIPTION
**Pull request by vladak**
This change makes it possible to configure bug/review links on per project level. Due to the way de-serialization of the `Project` objects works (setters are not called for unspecified properties), the getters have to be equipped with null checks and fall back to the global configuration. The code in `completeWithDefaults()` is used only when adding new projects in the indexer.

https://github.com/oracle/opengrok/wiki/Per-project-configuration will have to be updated.

Sample configuration:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<java version="11.0.4" class="java.beans.XMLDecoder">
 <object class="org.opengrok.indexer.configuration.Configuration" id="Configuration0">

  <void property="projectsEnabled">
   <boolean>true</boolean>
  </void>

   <void property="projects">
     <void method="put">
      <string>one</string>
      <object class="org.opengrok.indexer.configuration.Project">
       <void property="handleRenamedFiles">
        <boolean>true</boolean>
       </void>
       <void property="historyEnabled">
        <boolean>true</boolean>
       </void>
       <void property="indexed">
        <boolean>true</boolean>
       </void>
       <void property="name">
        <string>one</string>
       </void>
       <void property="path">
        <string>/one</string>
       </void>
       <void property="bugPage">
        <string>http://example.com/foobar</string>
       </void>
       <void property="mergeCommitsEnabled">
        <boolean>false</boolean>
       </void>
      </object>
     </void>
   </void>

 </object>
</java>
```

### CODE REVIEW

File: opengrok-indexer/src/main/java/org/opengrok/indexer/configuration/PatternUtil.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+public class PatternUtil {
+    private PatternUtil() {
+        // private to enforce static
+    }
+
+    /**
+     * A check if a pattern contains at least one pair of parentheses meaning
+     * that there is at least one capture group. This group must not be empty.
+     */
+    static final String PATTERN_SINGLE_GROUP = ".*\\([^\\)]+\\).*";
```
Comment (ahornace) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/configuration/PatternUtil.java:37]: Make these `private`?

File: opengrok-indexer/src/test/java/org/opengrok/indexer/configuration/ProjectTest.java [hunk 2]
```diff
@@ -143,6 +147,10 @@ public void testMergeProjects1() {
 
         assertEquals(env.getTabSize(), p1.getTabSize());
         assertEquals(env.isNavigateWindowEnabled(), p1.isNavigateWindowEnabled());
+        assertEquals(env.getBugPage(), p1.getBugPage());
```
Comment (ahornace) [inline opengrok-indexer/src/test/java/org/opengrok/indexer/configuration/ProjectTest.java:150]: Just FYI, recently found out about https://stackoverflow.com/questions/40796756/assertall-vs-multiple-assertions-in-junit5 which might be useful here? No need to change though

**Review by ahornace — APPROVED** [review]

File: opengrok-indexer/src/test/java/org/opengrok/indexer/configuration/ProjectTest.java (hunk 2 above)
Comment (vladak) [reply opengrok-indexer/src/test/java/org/opengrok/indexer/configuration/ProjectTest.java]: Nice tip !

**Comment by vladak** [comment]
When updating a configuration of a staging machine with the changes in order to set JDK bug links, I ran into an old issue with per project workflow (#2147). First thing I added the XML snippet with `bugPage`/`bugPattern` into read-only configuration and ran `opengrok-projadm` to merge it with the current configuration in the web app. This resulted in the `Project` objects disappearing from the configuration.

Luckily, thanks to the easily accessible ZFS snapshots (in the hidden by default `.zfs` directory) and the script wrapping `opengrok-projadm` that creates the snapshots whenever it is run, I was able to recover valid configuration quickly - otherwise I'd have to run the indexer for all projects.

When working with the per project workflow, the correct way is to modify the properties (once the project is added) like so:
```
curl -X PUT -d 'https://bugs.openjdk.java.net/browse/JDK-' http://localhost:8080/source/api/v1/projects/jdk8u/property/bugPage
curl -X PUT -d '([1-9][0-9]{6,7})' http://localhost:8080/source/api/v1/projects/jdk8u/property/bugPattern
```
and then grab the configuration from the web app and make it persistent.

**Pull request merged**